lib/new-admin/graphql/modules/authentication.js

```diff
@@ -0,0 +1,283 @@
+  return users.validateAuthToken(token, 'reset_password')
+    .then(r => {
+      if (!r.success) throw new authErrors.InvalidUrlError()
+      return { id: r.userID }
+    })
+    .catch(err => console.error(err))
+}
+
+const validateReset2FALink = token => {
+  if (!token) throw new authErrors.InvalidUrlError()
+  return users.validateAuthToken(token, 'reset_twofa')
+    .then(r => {
+      if (!r.success) throw new authErrors.InvalidUrlError()
+      return users.getUserById(r.userID)
+    })
+    .then(user => {
+      const secret = otplib.authenticator.generateSecret()
+      const otpauth = otplib.authenticator.keyuri(user.username, 'Lamassu Industries', secret)
+      return { user_id: user.id, secret, otpauth }
+    })
+    .catch(err => console.error(err))
+}
+
+const deleteSession = (sessionID, context) => {
+  if (sessionID === context.req.session.id) {
+    context.req.session.destroy()
+  }
+  return sessionManager.deleteSessionById(sessionID)
+}
+
+const login = (username, password) => {
+  return authenticateUser(username, password).then(user => {
+    if (!user) throw new authErrors.InvalidCredentialsError()
+    return users.getUserById(user.id).then(user => {
+      const twoFASecret = user.twofa_code
+      return twoFASecret ? 'INPUT2FA' : 'SETUP2FA'
+    })
+  })
+}
+
+const input2FA = (username, password, rememberMe, code, context) => {
+  return authenticateUser(username, password)
+    .then(user => {
+      if (!user) throw new authErrors.InvalidCredentialsError()
+      return users.getUserById(user.id)
+    })
+    .then(user => {
+      const secret = user.twofa_code
+      const isCodeValid = otplib.authenticator.verify({ token: code, secret: secret })
+      if (!isCodeValid) throw new authErrors.InvalidTwoFactorError()
+
+      const finalUser = { id: user.id, username: user.username, role: user.role }
+      context.req.session.user = finalUser
+      if (rememberMe) context.req.session.cookie.maxAge = REMEMBER_ME_AGE
+
+      return true
+    })
+}
+
+const setup2FA = (username, password, rememberMe, secret, codeConfirmation, context) => {
+  return authenticateUser(username, password)
+    .then(user => {
+      if (!user || !secret) throw new authErrors.InvalidCredentialsError()
+
+      const isCodeValid = otplib.authenticator.verify({ token: codeConfirmation, secret: secret })
+      if (!isCodeValid) throw new authErrors.InvalidTwoFactorError()
+
+      return users.getUserById(user.id)
+    })
+    .then(user => {
+      const finalUser = { id: user.id, username: user.username, role: user.role }
+      context.req.session.user = finalUser
+      if (rememberMe) context.req.session.cookie.maxAge = REMEMBER_ME_AGE
+
+      return users.save2FASecret(user.id, secret)
+    })
+    .then(() => true)
+}
+
+const changeUserRole = (code, id, newRole, context) => {
+  const action = (id, newRole) => users.changeUserRole(id, newRole)
+
+  if (!code) {
+    return action(id, newRole)
+  }
```

Comment:
Doesn't this mean that the user can still choose to not send the 2fa on the request and still perform privileged actions? Same thing for every other privileged action here.

Reply:
Fixed